### PR TITLE
Feature/project service

### PR DESCRIPTION
**Pull request by wkrzyzanowski**

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+                        <!--<goals>-->
+                            <!--<goal>build</goal>-->
+                        <!--</goals>-->
+                        <!--<configuration>-->
+                            <!--<repository>${project.groupId}/${project.artifactId}</repository>-->
+                            <!--<pullNewerImage>false</pullNewerImage>-->
+                        <!--</configuration>-->
+                    <!--</execution>-->
+                <!--</executions>-->
+            <!--</plugin>-->
```
Comment (lukaszfranczuk) [inline pom.xml]: please don't commit commented code - if it is not needed remove it

File: src/main/java/pl/betse/beontime/projectservice/controller/ClientController.java [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+    private final ClientMapper clientMapper;
+    private final ClientService clientService;
+
+    public ClientController(ClientMapper clientMapper, ClientService clientService) {
+        this.clientMapper = clientMapper;
+        this.clientService = clientService;
+    }
+
+    @GetMapping
+    public ResponseEntity<List<ClientBo>> ListOfAllClients() {
```
Comment (lukaszfranczuk) [inline src/main/java/pl/betse/beontime/projectservice/controller/ClientController.java]: we should not return BO objects to the clients

File: src/main/java/pl/betse/beontime/projectservice/controller/ClientController.java [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+    }
+
+    @PostMapping
+    public ResponseEntity createClient(@RequestBody ClientBody clientBody) {
+        ClientBo clientBo = clientService.addNewClient(clientMapper.mapClientBodyToClientBo(clientBody));
+        return ResponseEntity.ok(clientBo);
+    }
+
+    @PutMapping("/{guid}")
+    public ResponseEntity<ClientBo> updateClient(@PathVariable("guid") String guid, @RequestBody ClientBody clientBody) {
```
Comment (lukaszfranczuk) [inline src/main/java/pl/betse/beontime/projectservice/controller/ClientController.java]: this method doesn't return body, so it should be just ``ResponseEntity`` - without type

File: src/main/java/pl/betse/beontime/projectservice/controller/ClientController.java [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+        clientMapper.mapClientBoToClientBody(clientService.updateClient(guid, clientMapper.mapClientBodyToClientBo(clientBody)));
+        return ResponseEntity.ok().build();
+    }
+
+    @DeleteMapping("/{guid}")
```
Comment (lukaszfranczuk) [inline src/main/java/pl/betse/beontime/projectservice/controller/ClientController.java]: this method doesn't return body, so it should be just ``ResponseEntity`` - without type

File: src/main/java/pl/betse/beontime/projectservice/exception/ProjectAlreadyExistException.java [hunk 5]
```diff
@@ -0,0 +1,7 @@
+package pl.betse.beontime.projectservice.exception;
+
+import lombok.NoArgsConstructor;
+
+@NoArgsConstructor
```
Comment (lukaszfranczuk) [inline src/main/java/pl/betse/beontime/projectservice/exception/ProjectAlreadyExistException.java]: ``@NoArgsConstructor`` is redundant  - it is a default constructor in Java

File: src/main/java/pl/betse/beontime/projectservice/exception/ProjectNotFoundException.java [hunk 6]
```diff
@@ -0,0 +1,7 @@
+package pl.betse.beontime.projectservice.exception;
+
+import lombok.NoArgsConstructor;
+
+@NoArgsConstructor
```
Comment (lukaszfranczuk) [inline src/main/java/pl/betse/beontime/projectservice/exception/ProjectNotFoundException.java]: ``@NoArgsConstructor``is redundant - it is a default constructor in Java

**Review by lukaszfranczuk — CHANGES_REQUESTED** [review]

**Review by miles87 — DISMISSED** [review]

File: src/main/java/pl/betse/beontime/projectservice/exception/ClientAlreadyExistException.java [hunk 7]
```diff
@@ -0,0 +1,7 @@
+package pl.betse.beontime.projectservice.exception;
+
+import lombok.NoArgsConstructor;
+
+@NoArgsConstructor
```
Comment (lukaszfranczuk) [inline src/main/java/pl/betse/beontime/projectservice/exception/ClientAlreadyExistException.java]: ``@NoArgsConstructor`` is redundant  - this is default constructor in Java

File: src/main/java/pl/betse/beontime/projectservice/exception/ClientNotFoundException.java [hunk 8]
```diff
@@ -0,0 +1,7 @@
+package pl.betse.beontime.projectservice.exception;
+
+import lombok.NoArgsConstructor;
+
+@NoArgsConstructor
```
Comment (lukaszfranczuk) [inline src/main/java/pl/betse/beontime/projectservice/exception/ClientNotFoundException.java]: ``@NoArgsConstructor`` is redundant  - this is default constructor in Java

File: src/main/java/pl/betse/beontime/projectservice/mapper/ProjectMapper.java [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+            @Mapping(source = "departmentGuid", target = "departmentGuid")
+    })
+    ProjectBo mapProjectEntityToProjectBo(ProjectEntity projectEntity);
+
+    @Mappings({
+            @Mapping(source = "id", target = "guid", qualifiedByName = "mapGuid"),
+            @Mapping(source = "clientBo", target = "clientEntity"),
+            @Mapping(source = "departmentGuid", target = "departmentGuid")
+    }
+    )
```
Comment (lukaszfranczuk) [inline src/main/java/pl/betse/beontime/projectservice/mapper/ProjectMapper.java]: please use formatter before you push changes to the repository

File: src/main/java/pl/betse/beontime/projectservice/service/ProjectService.java [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+                .collect(Collectors.toList());
+    }
+
+    public ProjectBo addNewProject(ProjectBo projectBo) {
+        if (projectRepository.findByGuid(projectBo.getId()).isPresent()) {
+            throw new ProjectAlreadyExistException();
+        }
+
+        ProjectEntity projectEntity = projectMapper.mapProjectBoToProjectEntity(projectBo);
+        ClientEntity clientEntity = clientRepository.findByGuid(projectBo.getClientBo().getClientId()).orElseThrow(ClientNotFoundException::new);
```
Comment (lukaszfranczuk) [inline src/main/java/pl/betse/beontime/projectservice/service/ProjectService.java]: please format it properly

File: src/main/java/pl/betse/beontime/projectservice/service/ProjectService.java [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+        projectRepository.save(projectEntity);
+        return projectMapper.mapProjectEntityToProjectBo(projectEntity);
+    }
+
+    public ProjectBo updateProject(String guid, ProjectBo projectBo) {
+        ProjectEntity projectEntity = projectRepository.findByGuid(guid)
+                .orElseThrow(ProjectNotFoundException::new);
+        if (projectBo.getClientBo() == null) {
+            throw new ProjectNoClientException();
+        }
```
Comment (lukaszfranczuk) [inline src/main/java/pl/betse/beontime/projectservice/service/ProjectService.java]: this check should be done in cotroller - all input data should be validated there

**Review by lukaszfranczuk — CHANGES_REQUESTED** [review]

File: src/main/java/pl/betse/beontime/projectservice/controller/ClientController.java (hunk 2 above)
Comment (miles87) [reply src/main/java/pl/betse/beontime/projectservice/controller/ClientController.java]: resolved

**Review by lukaszfranczuk — APPROVED** [review]

**Pull request merged**